### ISSUE TITLE
Odd behavior with Distributed

### ISSUE THREAD

**Issue by daharn** [JuliaDynamics/DrWatson.jl#269]
When using `DrWatson` in combination with the `Distributed` package to do some multiprocessing tasks, I discovered some weird behavior. The whole operation appears to be very particular about the order of imports. But maybe my very poor understanding of how environments, packages, precompiling and distribution all work together is the reason.

MWE: Let's assume that the code lines I would like to distribute to worker processes make use of the directories defined by `@quickactivate` (`srcdir`, `datadir`, etc.) and imported packages (`Plots` in this MWE).

Somewhat expectedly, the following does not work:
```julia
using Distributed
addprocs(8)
using DrWatson 

@everywhere begin
    @quickactivate "TestEnv" # <- @quickactivate not defined on worker processes
    using Plots
    srcdir()
    p = plot

end
```
More surprisingly, the following does not work either 
```julia
using Distributed
addprocs(8)

@everywhere begin
    using DrWatson
    @quickactivate "TestEnv" # <- @quickactivate not defined on worker processes
    using Plots
    srcdir()
    p = plot

end
```
It appears like `using DrWatson` has no effect in an `@everywhere begin ... end` environment. This already seems buggy to me, but it gets weirder. The following will work, but only for a _single added process at a time_:
```julia
using Distributed
addprocs(1) # <- when adding more than one process, precompilation will most likely fail
@everywhere using DrWatson
@everywhere @quickactivate "TestEnv"

@everywhere begin
    using Plots
    srcdir()
    p = plot

end
```
The above code (almost?) always triggers precompilation of the `Plots` package (or any other package imported in the `@everywhere` section) _on all workers at the same time!_ This doesn't just spawn a ton of threads that can eat away all memory pretty quickly if the important packages have many dependencies, it also often results in access errors that will cause the whole Julia application to crash.

The only constellation that appears fully working to me is 
```julia
using Distributed
addprocs(8)
@everywhere using DrWatson

@everywhere begin
    @quickactivate "TestEnv"
    using Plots
    srcdir()
    p = plot

end
```
Can anyone explain to me what is happening here?

I'm working on a Windows machine with Julia 1.6.1 and DrWatson 2.0.5.

**Comment by Datseris** [comment]
I'm hoping @JonasIsensee  or @sebastianpech  or @tamasgal  have some idea, because I've never used `Distributed` in my life actually, so not sure how to tackle this...

**Comment by daharn** [comment]
I could move the question over to the discourse, but I could only reproduce this behavior with DrWatson so far. But it might just a well be something originating from `Distributed`.

**Comment by JonasIsensee** [comment]
Hi,

I've encountered this before as well and can shed some light onto the underlying problem:
The key problem is *macro expansion time*. This has nothing to do with either `DrWatson` or with `Distributed` but with the fact that `@quickactivate` is a macro.

Consider the following simple example:

```
julia> begin
           macro testmacro()
               println("this worked")
           end
           @testmacro
       end
ERROR: LoadError: UndefVarError: @testmacro not defined
in expression starting at REPL[1]:5

julia> begin
           macro testmacro()
               println("this worked")
           end
       end
@testmacro (macro with 1 method)

julia> begin
           @testmacro
       end
this worked
```


So what happens is that julia will try to first *compile* the whole `begin end` expression block and then execute it.
This usually works, but if the block defines AND uses a new macro, then this becomes impossible.

**Comment by Datseris** [comment]
So is the solution to have two `begin` blocks wrapped in `@everywhere`, like
```
@everywhere begin
using DrWatson
end

@everywhere begin
@quickactivate "la"
end
```
?

**Comment by JonasIsensee** [comment]
yes, but 

```
@everywhere using DrWatson
@everywhere begin
    @quickactivate "la"
    using MorePackages
 end
 ```
 will work as well. No need to wrap everything

**Comment by Datseris** [comment]
We should put this in the docstring of `@quickactivate` then

**Comment by daharn** [comment]
Thanks @JonasIsensee , that explains why my second example doesn't work. Macros are still a quite new concept to me.

But this doesn't explain the consistent triggering of the precompilation in example 3, does it?

**Comment by mbruna** [comment]
Thanks @daharn for bringing this up. I've been having exactly the same problem and is driving me crazy. Having done as suggested, my code runs fine with just one process, but still gives me errors when I try to run it with multiple workers:
```
From worker 3:    ┌ Warning: The call to compilecache failed to create a usable precompiled cache file for StatsBase [2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91]
      From worker 3:    │   exception = Required dependency SortingAlgorithms [a2af1166-a08f-5f64-846c-94a0d3cef48c] failed to load from a cache file.
      From worker 3:    └ @ Base loading.jl:1042
```
Any suggestions? Thanks!

**Comment by daharn** [comment]
Have  you tried my last example, where  `using DrWatson` gets its own `@everywhere` but `@quickactivate` is bundled together with the imports of the packages? For me, this also works for multiple processes at the time without triggering precompilation.

**Comment by daharn** [comment]
Is the issue regarding precompilation also tied to the effects of macro expansion? I couldn't make that out in the discussion so far. If so, would somebody care to elaborate?

**Comment by mbruna** [comment]
> Have you tried my last example, where `using DrWatson` gets its own `@everywhere` but `@quickactivate` is bundled together with the imports of the packages? For me, this also works for multiple processes at the time without triggering precompilation. I am still very curious why, though.

Yes, I did it that way, still errors. I haven't managed to figure out the correct order, as I have managed to get it to work some times but I can never reproduce it. The last time I thought I had understood how to get it to work was to first activate the DrWatson project and run the code once (only one iteration) to load all the packages in the parent process, and then addprocs and run it again. 

Could it make a difference that when I addprocs, they are in a set of different machines (all in the same network)?

**Comment by mbruna** [comment]
quick addition is that, if I re-run in for 4-5 times, each times the errors concern different packages and eventually it does work. No idea how to make it load everything at once.

**Comment by Datseris** [comment]
Are you sure this is related with DrWatson? Try not using DrWatson and just do `Pkg.activate("path/to/project.toml")` and have a single code block that `using` the packages you want.

If you can identify this is coming from sure from DrWatson, please open a new issue with a MWE, as the original discussion is resolved to my understanding.

**Comment by daharn** [comment]
Well, as the one who started the original discussion, I disagree. Maybe the confusion arose from the fact that I reported two separate(?) issues, of which only the first (macro expansion of `@quickactivate`) was resolved so far.

The second issue is related to the triggering of multiple seemingly unnecessary precompile actions on distributed workers, which is also seen by @mbruna .

I am relatively certain it is at least partly connected to `DrWatson`, because when I run
```julia
using Distributed
addprocs(8)

@everywhere begin
    import Pkg
    Pkg.activate(".") 
    using Plots
    p = plot
end
```
I don't get any issues.

However, I will gladly move this into a separate issue if it helps keeping everything in order.
